providers/implementations/signature/rsa_sig.c

```diff
@@ -216,13 +216,29 @@ static int rsa_pss_compute_saltlen(PROV_RSA_CTX *ctx)
      * Provide a way to use at most the digest length, so that the default does
      * not violate FIPS 186-4. */
     if (saltlen == RSA_PSS_SALTLEN_DIGEST) {
-        saltlen = EVP_MD_get_size(ctx->md);
+        if ((saltlen = EVP_MD_get_size(ctx->md)) <= 0) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_DIGEST);
+            return -1;
+        }
     } else if (saltlen == RSA_PSS_SALTLEN_AUTO_DIGEST_MAX) {
         saltlen = RSA_PSS_SALTLEN_MAX;
-        saltlenMax = EVP_MD_get_size(ctx->md);
+        if ((saltlenMax = EVP_MD_get_size(ctx->md)) <= 0) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_DIGEST);
+            return -1;
+        }
     }
     if (saltlen == RSA_PSS_SALTLEN_MAX || saltlen == RSA_PSS_SALTLEN_AUTO) {
-        saltlen = RSA_size(ctx->rsa) - EVP_MD_get_size(ctx->md) - 2;
+        int mdsize, rsasize;
+
+        if ((mdsize = EVP_MD_get_size(ctx->md)) <= 0) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_DIGEST);
+            return -1;
+        }
+        if ((rsasize = RSA_size(ctx->rsa)) <= 0 || rsasize < mdsize + 2) {
```

Comment:
Note that this isn't totally safe: with extreme values for `rsasize` and `mdsize` this test isn't sufficient. For example, a quick local test

```
$ cat overflow.c
#include <stdlib.h>
#include <stdio.h>

int main(int argc, char *argv[])
{
    int rsasize = atoi(argv[1]);
    int mdsize = atoi(argv[2]);

    if (rsasize < mdsize + 2) {
        printf("rsasize < mdsize + 2\n");
        exit(1);
    }

    int saltlen = rsasize - mdsize - 2;
    printf("saltlen = rsasize - mdsize - 2 = %d\n", saltlen);
}

$ make overflow
cc     overflow.c   -o overflow

$ ./overflow 2147483647 2147483647
saltlen = rsasize - mdsize - 2 = -2
```

A simple solution might be to limit the maximum values of `mdsize` and `rsasize` to something well within the limits of `int` but well outside the bounds of reasonableness for what they are - e.g. 1048576.

Reply:
`rsasize < 2 || rsasize - 2 < mdsize` works I think.
An alternative would be casting to unsigned or size_t with appropriate checks for negatives.

Reply:
mdsize+2 cannot really overflow unless the provider is malicious.

Reply:
fixup pushed